src/Avalonia.Controls/PullToRefresh/RefreshVisualizer.cs

```diff
@@ -0,0 +1,647 @@
+        public static readonly StyledProperty<PullDirection> PullDirectionProperty =
+            AvaloniaProperty.Register<RefreshVisualizer, PullDirection>(nameof(PullDirection), PullDirection.TopToBottom);
+
+        /// <summary>
+        /// Defines the <see cref="RefreshRequested"/> event.
+        /// </summary>
+        public static readonly RoutedEvent<RefreshRequestedEventArgs> RefreshRequestedEvent =
+            RoutedEvent.Register<RefreshVisualizer, RefreshRequestedEventArgs>(nameof(RefreshRequested), RoutingStrategies.Bubble);
+
+        /// <summary>
+        /// Defines the <see cref="RefreshVisualizerState"/> property.
+        /// </summary>
+        public static readonly DirectProperty<RefreshVisualizer, RefreshVisualizerState> RefreshVisualizerStateProperty =
+            AvaloniaProperty.RegisterDirect<RefreshVisualizer, RefreshVisualizerState>(nameof(RefreshVisualizerState),
+                s => s.RefreshVisualizerState);
+
+        /// <summary>
+        /// Defines the <see cref="Orientation"/> property.
+        /// </summary>
+        public static readonly DirectProperty<RefreshVisualizer, RefreshVisualizerOrientation> OrientationProperty =
+            AvaloniaProperty.RegisterDirect<RefreshVisualizer, RefreshVisualizerOrientation>(nameof(Orientation),
+                s => s.Orientation, (s, o) => s.Orientation = o);
+
+        /// <summary>
+        /// Defines the <see cref="RefreshInfoProvider"/> property.
+        /// </summary>
+        internal DirectProperty<RefreshVisualizer, RefreshInfoProvider?> RefreshInfoProviderProperty =
+            AvaloniaProperty.RegisterDirect<RefreshVisualizer, RefreshInfoProvider?>(nameof(RefreshInfoProvider),
+                s => s.RefreshInfoProvider, (s, o) => s.RefreshInfoProvider = o);
+
+        /// <summary>
+        /// Gets or sets a value that indicates the refresh state of the visualizer.
+        /// </summary>
+        protected RefreshVisualizerState RefreshVisualizerState
+        {
+            get
+            {
+                return _refreshVisualizerState;
+            }
+            private set
+            {
+                SetAndRaise(RefreshVisualizerStateProperty, ref _refreshVisualizerState, value);
+                UpdateContent();
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets a value that indicates the orientation of the visualizer.
+        /// </summary>
+        public RefreshVisualizerOrientation Orientation
+        {
+            get
+            {
+                return _orientation;
+            }
+            set
+            {
+                SetAndRaise(OrientationProperty, ref _orientation, value);
+            }
+        }
+
+        internal PullDirection PullDirection
+        {
+            get => GetValue(PullDirectionProperty);
+            set
+            {
+                SetValue(PullDirectionProperty, value);
+
+                OnOrientationChanged();
+
+                UpdateContent();
```

Comment:
Consider using OnPropertyChanged for the styled properties. As this setter won't be called, if you set value with SetValue call directly.

Reply:
Same isn't necessary for direct properties tho.